lib/extensions.js

```diff
@@ -80,37 +80,43 @@ module.exports = function(realmConstructor, context) {
     const { DefaultNetworkTransport } = require('realm-network-transport');
     realmConstructor._networkTransport = new DefaultNetworkTransport();
 
-    Object.defineProperty(realmConstructor.Object.prototype, "toJSON", {
-        value: function () {
-            const result = {}
-            Object.keys(this).forEach(p => result[p] = this[p]);
-            Object.keys(Object.getPrototypeOf(this)).forEach(p => result[p] = this[p]);
-            return result;
+    Object.defineProperty(realmConstructor.Collection.prototype, "toJSON", {
+        value: function (_, cache = new Map()) {
+            return this.map((item, index) => item.toJSON ? item.toJSON(index.toString(), cache) : item);
         },
 
         writable: true,
         configurable: true,
         enumerable: false
     });
 
-    Object.defineProperty(realmConstructor.Object.prototype, "keys", {
-        value: function () {
-            return Object.keys(this).concat(Object.keys(Object.getPrototypeOf(this)));
-        },
+    Object.defineProperty(realmConstructor.Object.prototype, "toJSON", {
+        value: function (_, cache = new Map()) {
+            // Construct a reference-id of table-name & objectId
+            const id = this.constructor.name + this._objectId();
+
+            // Check if current objectId has already processed, to keep object references the same.
+            const existing = cache.get(id);
+            if (existing) {
+                return existing;
+            }
 
-        writable: true,
-        configurable: true,
-        enumerable: false
-    });
+            // Create new result, and store in cache.
+            const result = {};
+            cache.set(id, result);
 
-    Object.defineProperty(realmConstructor.Object.prototype, "entries", {
-        value: function () {
-            let result = {};
-            for (const key in this) {
-                result[key] = this[key];
-            }
+            // Add the generated reference-id as a non-enumerable prop '$refId', for later exposure though e.g. Realm.JsonSerializationReplacer.
+            Object.defineProperty(result, "$refId", { value: id, configurable: true });
+
+            // Move all enumerable keys to result, triggering any specific toJSON implementation in the process.
+            Object.keys(this)
+                .concat(Object.keys(Object.getPrototypeOf(this)))
+                .forEach(key => {
+                    // TODO: check that we're dealing with a Realm type, before passing in cache?
+                    result[key] = this[key].toJSON ? this[key].toJSON(key, cache) : this[key];
```

Comment:
```suggestion
                    const value = this[key];
                    result[key] = value instanceof Realm.Collection || value instanceof Realm.Object
                        ? value.toJSON(key, cache)
                        : value;
```

Reply:
Wouldn't that be violating the contract of `toJSON`? If an object implements `toJSON`, shouldn't we call it recursively?

Reply:
I think we're fine. We call it recursively to ensure we can determine circular structures. I can test it.

Reply:
It depends on the interpretation of the spec of `toJSON`. I havn't been able to find a better spec for the `toJSON` implementation than https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#toJSON_behavior. Although its not stated explicitly `JSON.stringify` will call `toJSON` recursively on all the values of the object being returned from this method. If we for-example returned an object with a `Date` as a value, that date will be converted to a string by its `toJSON` even if we didn't explicitly called `toJSON` on it.

Reply:
I tried @kraenhansen's suggestion and `Date`s are still serialized to iso-strings & `ObjectId`s to hex-strings 👍.
It accommodates my own TODO (that was left in...), so I'll add this check.